recipe/meta.yaml

```
@@ -44,7 +44,7 @@ requirements:
     - cuda-nvml-dev    # [(cuda_compiler_version or "None").startswith("12")]
     - nccl
   run:
-    - {{ pin_compatible("ucx", max_pin="x.x.x") }}
+    - {{ pin_compatible("ucx") }}
```

Should we just not remove the whole `run:` section with the UCX pin and rely on the [UCX's run exports](https://github.com/conda-forge/ucx-split-feedstock/blob/main/recipe/meta.yaml#L20)?

---

This expands to
```yaml
- {{ pin_compatible('ucx', min_pin='x.x.x.x.x.x', max_pin='x') }}
```
so it'd have a lower bound pinning at `ucx>=1.18.1`. @pentschev if you want to also support 1.18.0 you can relax it to
```yaml
- {{ pin_compatible('ucx', min_pin='x.x', max_pin='x') }}
```

---

So we don't want to rely on run exports from UCX? I am not sure I understand why they are not sufficient.

---

So is this something that should in theory be fixed in UCX's run exports upstream?

---

Whatever we list here, it'd generate one constraint, and `run_exports` from UCX would add another. This is why Peter reported two constraints, and in terms of dependency resolution we'll be looking at the subset allowed by both. So to keep it simple while maintainable (we still want to know UCX is a run-time dependency) I suggest just list it:
 ```suggestion
     - ucx
```

---

I guess we could ignore run exports, but it seems to me like we should fix this upstream in UCX.

---

Just listing `ucx` should be fine, at least until `ucx=2.*` is available, in which case this release will be unbounded, right? Meaning, anyone could install the build after this PR even with `ucx=2.*`, which doesn't seem to be ideal but maybe it will be fine since in the far future it's unlikely this UCC build will still be used.

From https://github.com/conda-forge/staged-recipes/pull/25886#discussion_r1579768878, UCC should be compatible with major UCX versions, so presumably the ideal case would be to pin to `ucx=1.*`, right? I guess in that case what we would want is:

```suggestion
    - {{ pin_compatible("ucx", max_pin="x") }}
```

Please check the "conda-ness" of my comment, I'm not confident I understand pinning rules well enough.

---

Reading [this comment](https://github.com/openucx/ucc/issues/968#issuecomment-2077613218), I think we should be really careful about relaxing the UCX pin. Unless we confirm with UCX developers that UCS is ABI compatible between minor versions, we should not relax the pin. Have we confirmed UCS ABI compatibility?

---

Alright, I confirmed offline with the UCX development team about the comment above, UCS is not guaranteed to be backwards- nor forwards-compatible, even across commits. IOW, there are no compatibility guarantees even between patch releases. With that said, I think we really need to pin to the exact release version, and thus a UCC build against `ucx=1.18.0` is not compatible with a build against `ucx=1.18.1`, or vice-versa.

With that said, I think we need to build and run with a single, exact UCX version, and we also need to ensure UCC is rebuilt for every new UCX conda release. What is the correct way to achieve that?

---

I am no expert, but it seems that we have to modify UCX feedstock to add something like to the `bot:` section of `conda-forge.yml` in UCX:

```
  rebuild_on_version_change:
    - ucc
```

That should trigger a PR to UCC every time there is a new version of UCX. @leofang, can you confirm that this is the right solution?

---

No, there's no such thing as `rebuild_on_version_change`. Where was this from?

@conda-forge/ucx-split @conda-forge/openmpi @conda-forge/mpich @conda-forge/libnvshmem @conda-forge/mvapich @conda-forge/psmpi @conda-forge/holoscan @conda-forge/mcstas-suite @conda-forge/adios2 ([based on this GitHub search](https://github.com/search?q=org%3Aconda-forge+path%3Arecipe%2Fmeta.yaml+ucx&type=code)), we have a problem here. 

The global pinning on `ucx` is only major.minor
https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/e6606c79aea7c9232e72440e269f6b4531b499f6/recipe/conda_build_config.yaml#L985-L986
but according to @pentschev's investigation (with replies from a UCX core dev) it needs to be pinned at major.minor.patch, as one of its shared libraries `libucs` does not hold any guarantee on ABI compatibility unfortunately.

I hope there is a simple solution but I can't see any without rebuilding the whole world (that depends on `ucx`). I will be very happy to learn if I am wrong. I suggest to fix this in 5 steps:
1. We fix `run_exports` in ucx-split-feedstock with `max_pin='x.x.x'`
   - we may also have to mark past packages as broken, or patch their repo data
2. We merge a migrator in the global pinning feedstock to pin at `1.18.0`
3. I assume the bot will rebuild the whole world that depends on `ucx` with 1.18.0 
4. Once the migration is closed, we send another migrator (maybe the cf bot would do this automatically) to update the global pin to 1.18.1
5. I assume again the bot will rebuild the whole world with 1.18.1

After this is completed we'll have all packages that work with UCX either 1.18.0 or 1.18.1.

---

Sorry, `rebuild_on_version_change` was AI and not verify. Having looked at this a bit more, your plan seems the way to go. Probably would also be good to add some comment in UCX readme explaining the reason for the strict run export so that people don't try to "fix" it. Also, @pentschev, do we know if the ABI incompatibility is going to be the permanent state or if it will be stabilized in the future? If there is an issue for this, it would be helpful to watch it to change the strict pin when/if it is not necessary anymore.

---

Do we actually need to rebuild the world? This is exclusively a problem for libraries that directly use libucs, libraries that link directly to libucp for example wouldn't need to do that. So wouldn't rebuilding UCC in the current context suffice (assuming only rebuilding select libraries is possible)?

As for UCS stability, from what I could grasp there's currently no plan to make the ABI stable, as it is generally expected to be more of an implementation detail I believe, and UCC is somewhat of a special case since it's developed as part of the OpenUCX consortium. However, I've asked internally what would it take to make the UCS ABI stable, I think I'll only hear back on Monday though.

---

UCS may be used for rather trivial reasons. Taking MPICH as an example:
```
$ nm -D libmpi.so.12 | grep ucs_
                 U ucs_status_string
```
So, in this case, is a rebuild worth or not? Auditing every package will be time consuming and error prone.

Rant: If UCX cannot commit to any kind of ABI compatibility, then maybe they should not blindly expose non-stable symbols in `libucs.so`, and rather version/mangle them. On top of that, they do not bump the SOVERSION number but it stays at `.0`. Add to that the cyclic dependency between `libucs` and `libucm`... At this point, it is very hard for binary distributors of conda/wheel packages to do things right.

---

To clarify: the `ucs_*` types that are exposed via UCP _have guaranteed ABI stability_. What is not guaranteed is:

- Linking directly to `libucs.so`;
- Calling `ucs_*` functions.

None of the above are normally done if a library directly relies on UCP, for example.

> So, in this case, is a rebuild worth or not? Auditing every package will be time consuming and error prone.

To provide a direct answer to this question: no, provided there's no direct linkage to `libucs.so` and no `ucs_*` funciton calls.

---

> the `ucs_*` types that are exposed via UCP _have guaranteed ABI stability_.

Sorry, I'm admittedly ignorant about UCX. What that means, exactly?

> * Linking directly to `libucs.so`;

But if you use a symbol like `ucs_status_string`, which is defined in `libucs.so.0`, then your third-party binary should link to `libucs.so.0`. AFAICT, that's the recommended practice in Linux packaging.

---

`ucs_status_string` is a good point. I've asked about it to the UCX core dev team as well.

---

I think that since the UCX package exposes a non-stable component, it probably has to be strict on it's pin. It's hard to track through Conda who uses it and how.

---

If libucs and libucp have different abi stability, that could be addressed by splitting the outputs in ucx-split and giving them different pinning in run_exports.

---

@minrk yes, normally this is a way to go. But not in this case because for all projects that we've examined so far it seems `libucs` always ends up being linked explicitly (instead of being guarded behind the more stable `libucp`).

FWIW this is from Open MPI v5.0.6 (courtesy of @dalcinl):
```shell
$ nm -D /usr/lib64/openmpi/lib/libmpi.so | grep ucs_
                 U ucs_status_string
```

---

@leofang I did a quick experiment with the binaries of my MPICH wheels.  I removed manually the dependency on UC{T|S|M} from `libmpi.so.12`, thus leaving only `libucp.so.0` in the list of `DT_NEEDED`. After that hack, things still work, mpi4py is able to load and run.  Afterwards, I did the same with the Open MPI wheels, and mpi4py still works.

In both cases, the MPI library needs the `ucs_status_string`, wich is technically in UCS. Nonetheless, dlopening the MPI library opens `libucp.so.0` eventually, and that triggers the loading of `libucs.so.0`, and all UCX symbols are eventually resolved.  Moreover, dlopening the MPI libraries with `RTLD_NOW|RTLD_LOCAL|RTLD_DEEPBIND` still works.

The big big drawback of this hack is that the ELF patching has to be implemented in the consumers of UCX, and not in UCX itself. However, those who care maybe have a path forward to use less strict UCX pinning.

---

> But not in this case because for all projects that we've examined so far it seems `libucs` always ends up being linked explicitly (instead of being guarded behind the more stable `libucp`).

No, this is not right. So far the only project that explicitly links to `libucs` is UCC itself, thus the discussion happening here. All other projects (that I know of) link to `libucp`, then `libucp` links to `libucs` and resolves `ucs_status_string`. So yeah, I guess we could split outputs to resolve linking separately, but I think the only direct consumer of `libucs` will continue to be UCC.

> In both cases, the MPI library needs the ucs_status_string, wich is technically in UCS.

I've asked about `ucs_status_string` and the answer I've got is that it is not expected to change so we can generally assume that part of the API/ABI is stable. However, I'm still asking whether it would be possible to have stability guarantees better documented for packagers, that will certainly given more clarity to everyone, I do not have any responses or ETA on that though but will update it here when I hear back.

> The big big drawback of this hack is that the ELF patching has to be implemented in the consumers of UCX, and not in UCX itself. However, those who care maybe have a path forward to use less strict UCX pinning.

I don't think I follow that, what ELF patching has to be implemented by consumers of UCX? Why isn't dynamically linking to `libucp.so.0` sufficient?

---

>> (...) so far it seems `libucs` always ends up being linked explicitly (instead of being guarded behind the more stable `libucp`).
>
> No, this is not right. So far the only project that explicitly links to `libucs` is UCC itself, thus the discussion happening here.
>
> (...)
>
> I don't think I follow that, what ELF patching has to be implemented by consumers of UCX? Why isn't dynamically linking to `libucp.so.0` sufficient?

@pentschev this is what we've been saying, unfortunately all projects including both MPICH and Open MPI do dynamically link directly to `libucs` (it is listed as `DT_NEEDED`):
```shell
$ readelf -d lib/libmpi.so
...
 0x0000000000000001 (NEEDED)             Shared library: [libucs.so.0]
...
```
which can be examined with conda-forge `openmpi`:
```shell
$ wget https://anaconda.org/conda-forge/openmpi/5.0.7/download/linux-64/openmpi-5.0.7-hb85ec53_100.conda
$ cph extract openmpi-5.0.7-hb85ec53_100.conda
$ cd openmpi-5.0.7-hb85ec53_100/
$ readelf -d lib/libmpi.so
```

So, what @dalcinl was saying is that he had to use patchelf or lief to patch the ELF file (in this case, `libmpi.so.XX`) to remove this `DT_NEEDED` entry. And this must be done to UCX-consuming libraries, not UCX itself.

---

Summarizing my understanding. As of now we have two solutions:
1. "Rebuild the whole world" ([link](https://github.com/conda-forge/ucc-feedstock/pull/25#discussion_r2080520612))
2. Split the `ucx` package into `ucs` and "non-UCS" (as per @minrk [here](https://github.com/conda-forge/ucc-feedstock/pull/25#discussion_r2082837823)), and then `ucx`-consuming libraries patch away their link to `ucs` (as per @dalcinl [here](https://github.com/conda-forge/ucc-feedstock/pull/25#discussion_r2087255557))

btw I've also pinged a few build experts on our side to help take a look and explore alternatives. But really if we can get commitment from the UCX team (with which @pentschev has been in touch) it'd be the ultimate solution...

---

An extra thing we could ask the UCX team: Add a function `ucp_status_string` to `libucp.so` that wraps `ucs_status_string`, this way MPI and others can call `ucp_status_string`, and this way UCP consumers  can strictly depend ONLY on `libucp.so`.

---

It would be great to get a `ucp_status_string` if that removes the only symbol directly used. From my reading I fully agree, I had thought the UCX libraries provided forward compatibility at a minimum, it seems like we would need a `ucc` build against any `ucx` release, and splitting the ABI stable and less ABI stable seems like the safest way to ensure `ucx`-consuming packages only depend on the more ABI stable. It would be great to fully document the ABI guarantees or lack thereof.

---

I've asked the UCX team if they could at least guarantee compatibility between patch versions, which is basically the status quo of the UCX package. Also, looking at the code in the OpenMPI repo, it seems that there is potentially more dependency between UCS and OpenMPI than just the `ucp_status_string`, but maybe that depends on how OpenMPI is configured.

---

> this is what we've been saying, unfortunately all projects including both MPICH and Open MPI do end up dynamically linking directly to `libucs` (it is listed as `DT_NEEDED`):

Ah, ok, my apologies, I didn't realize OpenMPI/MPICH were indeed linking directly, in that case I agree we have no better options at the moment rather than rebuilding them for every new UCX version.

I also went ahead and opened two issues to discuss/track the issues we've discussed here:
- Request public documentation for ABI guarantees: https://github.com/openucx/ucx/issues/10687
- Request for ABI compatibility for all UCX submodules across releases of a same minor version: https://github.com/openucx/ucx/issues/10688

---

it might be relevant to note that ucx _explicitly instructs_ dependents to directly link libucs via its `pkg-config` outputs, etc. If they don't want folks to link libucs directly, they should probably change [how they instruct packages to link ucx](https://openucx.readthedocs.io/en/master/faq.html#build-user-application-with-ucx).

```
$ pkg-config --libs ucx
-L$PREFIX/lib -lucp -luct -lucs -lucm
```

That's not necessarily wrong, since all of those together do form ucx, but there is no similar pkg-config for just ucp, etc. The distinction is only really relevant if the ABI stability is not consistent across components.

---

Yes, I think linking is indeed not necessarily wrong, what is confusing is the ABI guarantees, given according to latest discussions not all UCX submodules promise the same ABI stability making it confusing to know what to do in cases such as the present one.

---

if ucx isn't interested and keeps ucs with no patch-level commitment, while the others _do_, splitting libucs into its own output with its own run_exports will allow us to address this via the existing conda-forge-pinning mechanism, as the ucx pin can be `x.x` and the ucs pin can be `x.x.x`. All the conda-forge automations will work once packages that link libucs explicitly depend on libucs in host.

But as we're discovering, it may be that ~all packages that depend on ucx link libucs directly, in which case we might as well make ucx's own run_exports pin x.x.x, which is far simpler.

@leofang mentioned "rebuilding the world", but on conda-forge, that's 12 packages:

```
mamba repoquery whoneeds -c conda-forge/linux-64 ucx | awk '{print $1}' |  uniq
---
adios2
libadios2
libarrow
libarrow-flight
libholoscan
libnvshmem3
mcstas
mcxtrace
mpich
mvapich
openmpi
psmpi
ucc
```

That's actually slightly _over_ stating it, since only older builds of libarrow/libarrow-flight link ucx, current builds do not.

So far, every package I've checked links libucs directly (you can check for `Needed DSO lib/libucs.so.0` in build logs), which means every single package that currently depends on ucx would also get a new dependency on libucs for the tighter pin. Given that, the simplest option may be to take a least common denominator approach and treat ucx abi as its least stable component, and rebuild its relatively few dependents on patch releases of ucx instead of minor releases. At least until the ucx project documents an ABI plan.

The only approach that's not easily automateable is having no package to represent libucs with 'x.x.x' run_exports, but rebuilding some packages _more_ often than others. i.e. ucc can pin ucx more strictly than everyone else easily enough, with the only hitch that it won't get automated rebuild PRs when there's a ucx patch release. A manual rerender without other changes will get it, though.

---

> @leofang mentioned "rebuilding the world", but on conda-forge, that's 12 packages:
> (...) 
> That's actually slightly _over_ stating it, since only older builds of libarrow/libarrow-flight link ucx, current builds do not.

Well, I did use quotation marks and [ping maintainers of all impacted packages](https://github.com/conda-forge/ucc-feedstock/pull/25#discussion_r2080520612), so I assume my "overstatement" was clear enough regarding the scope of rebuilding 🙂 The "world" that needs high-performance communication libraries is small, after all...

I like @minrk's analysis. Sounds like we're in favor of plan 1 (rebuilding + fixing `ucx`'s `run_exports` to `x.x.x`). Do [the steps that I outlined above](https://github.com/conda-forge/ucc-feedstock/pull/25#discussion_r2080520612) sound reasonable? In particular, how do we fix existing packages' dependency on `ucx`? I don't think we have enough information to patch their repodata?  

If we are cool with this plan, can anyone help execute it?

---

> In particular, how do we fix existing packages' dependency on ucx?

If we want to pin packages as if they were built with `x.x.x` upper bound that we will add to ucx run_exports for future builds, that is doable with repodata patches `tighten_depends`. I think we may be able to know that we don't need to worry about this for pre-1.18 builds, too, so it's only those depending on 1.18.x that would get a patch.

So if tightening ucx run_exports to x.x.x and applying that retroactively and in the future is 3 operations:

1. tighten run_exports in ucx feedstock
2. add 3-part version pinning to conda-forge-pinning so packages will get rebuilds on every x.y.z release, not just x.y, and a _migrator_ if we need to make sure everything is built with the latest patch version
3. repodata patch with `tighten_depends: x.x.x` for those depending on ucx 1.18.x (effectively applying the x.x.x retroactively)

---

> we may also have to mark past packages as broken, or patch their repo data

IMO, given we have not seen any such reports that any existing packages may have hit any potential ABI breakage, I would say we just leave old packages "as is". After all, we discovered this "by accident", thanks to https://github.com/conda-forge/ucc-feedstock/pull/25#discussion_r2078617699.

> I think we may be able to know that we don't need to worry about this for pre-1.18 builds, too, so it's only those depending on 1.18.x that would get a patch.
> 3\. repodata patch with `tighten_depends: x.x.x` for those depending on ucx 1.18.x (effectively applying the x.x.x retroactively)

If patching only 1.18.x's repodata suffices for the most part, I'm also in favor of leaving older builds untouched.

---

> that is doable with repodata patches `tighten_depends`.

Good to know `tighten_depends` exists -- I wasn't sure if the build-time version is stored. Sounds like it is, then this should work just fine! 

I am fine if we only patch 1.18.x. I don't have concerns here.

---

Sounds reasonable to me, I've applied both your suggestions @jakirkham . Would be good if everyone could have another look and let us know if there are still open issues after those.